src/components/InngestStepTools.ts

```diff
@@ -0,0 +1,473 @@
+      readingFromStack = false;
+
+      const submitOp: Parameters<
+        NonNullable<Parameters<typeof createTool>[1]>
+      >[0]["submitOp"] = (...args) => {
+        state.nextOp = new Promise((resolve) => resolve(args[0])).then(
+          (data) => ({
+            ...opId,
+            ...(args.length ? { data } : {}),
+          })
+        );
+      };
+
+      /**
+       * If we've been passed a custom function to run in response to user data,
+       * run that now. That function will then be responsible for submitting an
+       * op.
+       *
+       * If we haven't been given this, simply submit the op.
+       */
+      if (fn) {
+        fn({ submitOp }, ...args);
+      } else {
+        submitOp();
+      }
+
+      /**
+       * If we've run the tool and it hasn't submitted an op, then we should
+       * throw. This is exceedingly unexpected and indicates that a tool has
+       * a bug.
+       */
+      if (!state.nextOp) {
+        throw new Error("No operation was submitted by a tool");
+      }
+
+      /**
+       * If we're here, we've attempted to use a tool and should therefore throw
+       * an error to stop the function from running.
+       */
+      throw new StepFlowInterrupt();
+    }) as T;
+  };
+
+  const state: {
+    nextOp: Promise<HashedOp> | undefined;
+  } = {
+    nextOp: undefined,
+  };
+
+  /**
+   * Define the set of tools the user has access to for their step functions.
+   *
+   * Each key is the function name and is expected to run `createTool` and pass
+   * a generic type for that function as it will appear in the user's code.
+   */
+  const tools = {
+    /**
+     * Wait for a particular event to be received before continuing. When the
+     * event is received, it will be returned.
+     *
+     * You can also provide options to control the particular event that is
+     * received, for example to ensure that a user ID matches between two
+     * events, or to only wait a maximum amount of time before giving up and
+     * returning `null` instead of any event data.
+     */
+    waitForEvent: createTool<
+      <
+        IncomingEvent extends keyof Events,
+        Opts extends WaitForEventOpts<
+          Events[TriggeringEvent],
+          Events[IncomingEvent]
+        >
+      >(
+        event: IncomingEvent,
+        opts?: Opts
+      ) => Opts["timeout"] extends string
+        ? Opts["timeout"] extends ""
+          ? Events[IncomingEvent]
+          : Events[IncomingEvent] | null
+        : Events[IncomingEvent]
+    >(
+      (
+        /**
+         * The event name to wait for.
+         */
+        event,
+
+        /**
+         * Options to control the event we're waiting for.
+         */
+        opts
+      ) => {
+        const matchOpts: { ttl?: string; match?: string } = {};
+
+        if (opts?.timeout) {
+          matchOpts.ttl =
+            typeof opts.timeout === "string"
+              ? opts.timeout
+              : dateToTimeStr(opts.timeout);
```

Comment:
@djfarrelly Not sure what you mean by "_this timeout could float_," but this function will run, compare the date given to now (`new Date()`), create a time string from it (e.g. `"1h15m10s"`), then send that time string to Inngest and wait to be called again.

There's no local timeout that happens if that's what you were getting at? Sleepy Saturday morning here so forgive me if I've misunderstood. 😄

Reply:
Sorry, my thought wasn't very clear. I was trying to think about ways that this could produce unintended results, but I think it's not something we can handle in code here, the dev just needs to understand how it works.

The idea in my head is that we can accept a `Date`, which feels exact, but then we convert to something that is relative (ttl), so I was trying to think in what cases (retries, etc.) where a user providing a Date could expect something to be more exact (e.g. scheduling) than it will real be. I could see if a user decides to use a timestamp from the event payload to pass to the TTL, retries may fail depending on the data, but that may just be something devs need to learn to handle in code, not something we could necessarily optimize for in the library. We should definitely add pointers to out docs for using this tool :)

As a side note: I think in our examples, we should use [`dayjs`](https://day.js.org/) to show this could be used.

Reply:
Gotchya. Aye, I think you, @djfarrelly, and @tonyhb have converged on this idea! I had been matching the TTL field in the executor for the moment, but dates everywhere would be much nicer, even under the hood.

If we move to pass full timestamps to the executor after this, we don't have to worry so much about confusing conversions. We still have clock skew problems technically, but aye there's not too much we can do to remedy that other than potentially mentioning somewhere that any relative times passed are relative _to the machine running the code_.